cardano-client-demo/MainPipeline.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -freduction-depth=0 #-}
+
+import Cardano.Api
+    ( Block(..),
+      BlockHeader(BlockHeader),
+      ChainPoint,
+      ChainTip(ChainTip),
+      NetworkId(Mainnet),
+      BlockNo(BlockNo) )
+import qualified Cardano.Api.IPC as IPC
+import qualified Cardano.Chain.Slotting as Byron (EpochSlots(..))
+
+-- TODO: Export this via Cardano.Api
+import Ouroboros.Network.Protocol.ChainSync.ClientPipelined
+    ( ChainSyncClientPipelined(ChainSyncClientPipelined),
+      ClientPipelinedStIdle(SendMsgDone, SendMsgRequestNextPipelined,
+                            CollectResponse),
+      ClientStNext(..) )
+import Network.TypedProtocol.Pipelined
+    ( N(..), Nat(..), natToInt, unsafeIntToNat )
+
+import Control.Monad (when)
+import Data.Kind
+import Data.Proxy
+import Data.Time
+import qualified GHC.TypeLits as GHC
+import System.Environment (getArgs)
+import System.FilePath ((</>))
+
+-- | Connects to a local cardano node, requests the blocks and prints out the
+-- number of transactions. To run this, you must first start a local node e.g.:
+--
+--     $ cabal run cardano-node:cardano-node -- run \
+--        --config        configuration/cardano/mainnet-config.json \
+--        --topology      configuration/cardano/mainnet-topology.json \
+--        --database-path db \
+--        --socket-path   db/node.sock \
+--        --host-addr     127.0.0.1 \
+--        --port          3001 \
+--
+-- Then run this with the path to the directory containing node.sock:
+--
+--     $ cabal run cardano-client-demo-pipeline -- db
+--
+main :: IO ()
+main = do
```

Comment:
We're probably going to have a series of demos showing different and more features. Perhaps we should not have called the first one `Main.hs` :grin:, and establish a naming scheme for the demo variants.